namespace Infrastructure.Exceptions;

/// <summary>
/// This exception is thrown if an entity excepted to be found but not found.
/// </summary>
[Serializable]
public class EntityNotFoundException : InfraException
{
    private const string _errorFormat = "There is no such an entity type '{0}' with id: '{1}'";

    /// <summary>
    /// Creates a new <see cref="EntityNotFoundException"/> object.
    /// </summary>
    public EntityNotFoundException(Type entityType, object id)
        : base(string.Format(_errorFormat, entityType.Name, id))
    { }

    /// <summary>
    /// Creates a new <see cref="EntityNotFoundException"/> object.
    /// </summary>
    public EntityNotFoundException(Type entityType, object id, Exception innerException)
        : base(string.Format(_errorFormat, entityType.Name, id), innerException)
    { }

    /// <summary>
    /// Creates a new <see cref="EntityNotFoundException"/> object.
    /// </summary>
    /// <param name="message">Exception message</param>
    public EntityNotFoundException(string message)
        : base(message)
    { }

    /// <summary>
    /// Creates a new <see cref="EntityNotFoundException"/> object.
    /// </summary>
    /// <param name="message">Exception message</param>
    /// <param name="innerException">Inner exception</param>
    public EntityNotFoundException(string message, Exception innerException)
        : base(message, innerException)
    { }
}

